File: backend/app/api/routers/drive_connector.py
"""Drive connector admin 라우트 — WORK-003 (SPEC-004 API Contract).

- GET  /admin/drive-connector            : connector 상태 조회 (admin)
- POST /admin/drive-connector/watch      : watch channel 생성/갱신 (admin)
- POST /admin/drive-connector/sync/retry : 실패 sync 재처리, 멱등 (admin)
- GET  /admin/drive-sync-events          : sync 감사 목록, 최신순 (admin)

에러봉투 {detail:{error_code,message}} — SPEC-004 Case Matrix.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from app.api.deps import get_drive_sync_service, require_admin_only
from app.api.errors import SPEC003_004_ERRORS, spec003_004_http_error
from app.dtos.user import UserDTO
from app.schemas.drive_sync import (
    ConnectorStatusOut,
    SyncEventListResponse,
    SyncEventOut,
    SyncRetryResponse,
)
from app.services.drive_sync import ConnectorStatus, DriveSyncService

router = APIRouter(tags=["drive-connector"])


def _status_out(status: ConnectorStatus) -> ConnectorStatusOut:
    return ConnectorStatusOut(
        status=status.status,  # type: ignore[arg-type]
        scope=status.scope,
        selected_folder_id=status.selected_folder_id,
        selected_folder_name=status.selected_folder_name,
        watch_channel_id=status.watch_channel_id,
        watch_expires_at=status.watch_expires_at,
        last_sync_at=status.last_sync_at,
        last_error=status.last_error,
        page_token=status.page_token,
    )


@router.get("/admin/drive-connector", response_model=ConnectorStatusOut)
async def get_drive_connector(
    _admin: UserDTO = Depends(require_admin_only),
    service: DriveSyncService = Depends(get_drive_sync_service),
) -> ConnectorStatusOut:
    return _status_out(await service.connector_status())


@router.post("/admin/drive-connector/watch", response_model=ConnectorStatusOut)
async def register_drive_watch(
    _admin: UserDTO = Depends(require_admin_only),
    service: DriveSyncService = Depends(get_drive_sync_service),
) -> ConnectorStatusOut:
    try:
        return _status_out(await service.register_watch())
    except SPEC003_004_ERRORS as exc:
        raise spec003_004_http_error(exc)


@router.post("/admin/drive-connector/sync/retry", response_model=SyncRetryResponse)
async def retry_drive_sync(
    _admin: UserDTO = Depends(require_admin_only),
    service: DriveSyncService = Depends(get_drive_sync_service),
) -> SyncRetryResponse:
    try:
        summary = await service.retry_sync()
    except SPEC003_004_ERRORS as exc:
        raise spec003_004_http_error(exc)
    return SyncRetryResponse(
        processed=summary.processed,
        new_documents=summary.new_documents,
        updated_documents=summary.updated_documents,
        unavailable_documents=summary.unavailable_documents,
        skipped=summary.skipped,
        failed=summary.failed,
        page_token=summary.page_token,
    )


@router.get("/admin/drive-sync-events", response_model=SyncEventListResponse)
async def list_drive_sync_events(
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    _admin: UserDTO = Depends(require_admin_only),
    service: DriveSyncService = Depends(get_drive_sync_service),
) -> SyncEventListResponse:
    events, total = await service.list_events(limit=limit, offset=offset)
    return SyncEventListResponse(
        events=[
            SyncEventOut(
                id=e.id,
                event_type=e.event_type,  # type: ignore[arg-type]
                drive_file_id=e.drive_file_id,
                document_id=e.document_id,
                occurred_at=e.occurred_at,
                result=e.result,  # type: ignore[arg-type]
                message=e.message,
            )
            for e in events
        ],
        total=total,
        limit=limit,
        offset=offset,
    )

File: frontend/lib/api/driveConnector.ts
// SPEC-004 Drive connector/sync typed client — WORK-003 Phase 4.
// backend: app/schemas/drive_sync.py · app/schemas/documents.py 계약 기준.

import { api, ApiError } from "./client";

// ── types ────────────────────────────────────────────────────────────────────

export type ConnectorState =
  | "connected"
  | "disconnected"
  | "watch_expiring"
  | "error";

export interface ConnectorStatus {
  status: ConnectorState;
  scope: string;
  selected_folder_id: string | null;
  selected_folder_name: string | null;
  watch_channel_id: string | null;
  watch_expires_at: string | null;
  last_sync_at: string | null;
  last_error: string | null;
  page_token: string | null;
}

export interface SyncRetryResult {
  processed: number;
  new_documents: number;
  updated_documents: number;
  unavailable_documents: number;
  skipped: number;
  failed: number;
  page_token: string | null;
}

export type SyncEventType =
  | "webhook_received"
  | "changes_listed"
  | "document_upserted"
  | "document_unavailable"
  | "candidate_staled"
  | "reanalysis_enqueued"
  | "sync_failed";

export interface SyncEvent {
  id: number;
  event_type: SyncEventType;
  drive_file_id: string | null;
  document_id: number | null;
  occurred_at: string;
  result: "success" | "skipped" | "failed";
  message: string | null;
}

export interface SyncEventListResponse {
  events: SyncEvent[];
  total: number;
  limit: number;
  offset: number;
}

// ── SPEC-004 Case Matrix 프론트 출력 ─────────────────────────────────────────

const SPEC004_MESSAGES: Record<string, string> = {
  DRIVE_CONNECTOR_NOT_CONFIGURED: "Drive 연동 설정이 필요합니다.",
  DRIVE_FOLDER_NOT_CONFIGURED: "감시 폴더를 설정하세요.",
  DRIVE_WATCH_EXPIRED: "Drive watch 갱신이 필요합니다.",
  DRIVE_WEBHOOK_INVALID: "유효하지 않은 Drive 알림입니다.",
  DRIVE_CHANGES_FAILED: "Drive 변경 조회에 실패했습니다.",
  DRIVE_FILE_OUT_OF_SCOPE: "감시 범위 밖 문서입니다.",
  FORBIDDEN_ADMIN_ONLY: "관리자만 사용할 수 있습니다.",
};

/** ApiError → 사용자에게 보여줄 한국어 카피 (SPEC-004 Case Matrix '프론트 출력'). */
export function spec004Message(err: unknown, fallback: string): string {
  if (err instanceof ApiError && err.errorCode && SPEC004_MESSAGES[err.errorCode]) {
    return SPEC004_MESSAGES[err.errorCode];
  }
  return fallback;
}

// ── calls ────────────────────────────────────────────────────────────────────

export function fetchConnectorStatus(): Promise<ConnectorStatus> {
  return api<ConnectorStatus>("/admin/drive-connector");
}

export function registerWatch(): Promise<ConnectorStatus> {
  return api<ConnectorStatus>("/admin/drive-connector/watch", { method: "POST" });
}

export function retrySync(): Promise<SyncRetryResult> {
  return api<SyncRetryResult>("/admin/drive-connector/sync/retry", {
    method: "POST",
  });
}

export function fetchSyncEvents(
  limit = 30,
  offset = 0,
): Promise<SyncEventListResponse> {
  return api<SyncEventListResponse>(
    `/admin/drive-sync-events?limit=${limit}&offset=${offset}`,
  );
}
